pkg/deploy/clientset.go

```diff
@@ -0,0 +1,90 @@
+/*
+ * @license
+ * Copyright 2023 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package deploy
+
+import (
+	"context"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/client"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/client/automation"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/client/dtclient"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/manifest"
+)
+
+type clientSet struct {
+	// dtClient is the client capable of updating or creating settings and classic configs
+	dtClient dtclient.Client
+	// autClient is the client capable of updating or creating automation API configs
+	autClient automationClient
+}
+
+func (cs clientSet) automation() automationClient {
+	return cs.autClient
+}
+
+func (cs clientSet) settings() dtclient.Client {
+	return cs.dtClient
+}
+
+func (cs clientSet) classic() dtclient.Client {
+	return cs.dtClient
+}
+
+func NewClientSet(dtClient dtclient.Client, autClient *automation.Client) *clientSet {
+	return &clientSet{
+		dtClient:  dtClient,
+		autClient: autClient,
+	}
+}
+
+func CreateClientSet(url string, auth manifest.Auth, dryRun bool, opts ...func(dynatraceClient *dtclient.DynatraceClient)) (*clientSet, error) {
+	var dtClient dtclient.Client
+	var autClient automationClient
+	var err error
+	if dryRun {
+		return &clientSet{
+			dtClient:  &dtclient.DummyClient{},
+			autClient: &dummyAutomationClient{}}, nil
+	}
+	switch {
+	case dryRun:
+		dtClient = &dtclient.DummyClient{}
+		autClient = &dummyAutomationClient{}
+	case auth.OAuth == nil:
+		dtClient, err = dtclient.NewClassicClient(url, auth.Token.Value, opts...)
+		autClient = &dummyAutomationClient{}
+	case auth.OAuth != nil:
+		oauthCredentials := client.OauthCredentials{
+			ClientID:     auth.OAuth.ClientID.Value,
+			ClientSecret: auth.OAuth.ClientSecret.Value,
+			TokenURL:     auth.OAuth.GetTokenEndpointValue(),
+		}
+		dtClient, err = dtclient.NewPlatformClient(url, auth.Token.Value, oauthCredentials, opts...)
+		autClient = automation.NewClient(url, client.NewOAuthClient(context.TODO(), oauthCredentials))
+	default:
+		return nil, fmt.Errorf("unable to create authorizing HTTP Client for environment %s - no oauth credentials given", url)
```

Comment:
I think this message is at least partially wrong - "no oauth credentials given" would be the `case auth.OAuth == nil` - if we reach the default case something else went wrong